kafka_utils/kafka_consumer_manager/util.py

```diff
@@ -169,44 +168,84 @@ def __init__(self, kafka_config):
+        self.active_partitions = {}
+        self._finished = False
 
     def read_group(self, group_id):
         partition_count = get_offset_topic_partition_count(self.kafka_config)
         partition = get_group_partition(group_id, partition_count)
         return self.read_groups(partition).get(group_id, [])
 
     def read_groups(self, partition=None):
-        self.log.info("Kafka consumer running")
         self.consumer = KafkaConsumer(
             group_id='offset_monitoring_consumer',
             bootstrap_servers=self.kafka_config.broker_list,
             auto_offset_reset='earliest',
             enable_auto_commit=False,
-            consumer_timeout_ms=3000,
+            consumer_timeout_ms=30000,
+            fetch_max_wait_ms=2000,
+            max_partition_fetch_bytes=10 * 1024 * 1024,  # 10MB
         )
 
         if partition is not None:
-            self.consumer.assign([TopicPartition(CONSUMER_OFFSET_TOPIC, partition)])
+            self.active_partitions = {
+                partition: TopicPartition(CONSUMER_OFFSET_TOPIC, partition),
+            }
         else:
-            self.consumer.assign(
-                [
-                    TopicPartition(CONSUMER_OFFSET_TOPIC, p)
-                    for p in self.consumer.partitions_for_topic(CONSUMER_OFFSET_TOPIC)
-                ]
-            )
-
-        self.log.info("Consumer ready")
-        self.watermarks = self.get_current_watermarks(partition)
+            self.active_partitions = {
+                p: TopicPartition(CONSUMER_OFFSET_TOPIC, p)
+                for p in self.consumer.partitions_for_topic(CONSUMER_OFFSET_TOPIC)
+            }
+        self.watermarks = self.get_current_watermarks(self.active_partitions.values())
+        # Active partitions are not empty. Remove the empty ones.
+        self.active_partitions = {
+            p: tp for p, tp in self.active_partitions.items()
+            if tp.partition in self.watermarks and
+            self.watermarks[tp.partition].highmark > 0 and
+            self.watermarks[tp.partition].highmark > self.watermarks[tp.partition].lowmark
+        }
+        # Cannot consume if there are no active partitions
+        if not self.active_partitions:
+            return {}
+
+        self.consumer.assign(self.active_partitions.values())
+        self.log.info("Consuming from %s", self.active_partitions)
         while not self.finished():
-            message = self.consumer.next()
-            if message is None:
-                raise ConsumerTimeout("Timed out fetching messages from kafka")
-            max_offset = self.get_max_offset(message.partition)
-            if message.offset >= max_offset - 1:
-                self.finished_partitions.add(message.partition)
+            try:
+                message = self.consumer.next()
+            except StopIteration:
+                continue
+            if message.offset >= self.watermarks[message.partition].highmark - 1:
```

Comment:
It may be worthwhile to have a comment here explaining that highwatermark is the next future message's offset